django-stubs/utils/functional.pyi

```diff
@@ -92,9 +92,9 @@ _Get = TypeVar("_Get", covariant=True)
 
 class classproperty(Generic[_Get]):
     fget: Callable[[Self], _Get] | None
-    def __init__(self: Self, method: Callable[[Self], _Get] | None = ...) -> None: ...
-    def __get__(self: Self, instance: Self | None, cls: type[Self] = ...) -> _Get: ...
-    def getter(self: Self, method: Callable[[Self], _Get]) -> classproperty[_Get]: ...
+    def __init__(self, method: Callable[[Self], _Get] | None = ...) -> None: ...
```

Comment:
Now `Self` seems unbound. It can be `Any`

Reply:
I don't actually understand how this mechanism works 😆 I just reverted the unwanted changes and verified via the test that it works.

If I add `self: Self` here, it no longer works.

Reply:
mypy binds variables in two cases:
1. If they are in `Generic[]` definition, like we have here with `_Get`
2. If it is used more than once in a function: `def some(a: A) -> A: return a`

Here `Self` is unbound and should be changed to `Any` with a comment, I guess.

Reply:
maybe it could be done like in [python/typeshed stdlib/builtins.pyi classmethod](https://github.com/python/typeshed/blob/307dd08a6e09f800cb239d37f0eb475345c7c67c/stdlib/builtins.pyi#L128-L139)
```python
_R_co = TypeVar("_R_co", covariant=True)
...
class classmethod(Generic[_R_co]):
    @property
    def __func__(self) -> Callable[..., _R_co]: ...
    @property
    def __isabstractmethod__(self) -> bool: ...
    def __init__(self: classmethod[_R_co], __f: Callable[..., _R_co]) -> None: ...
    def __get__(self, __obj: _T, __type: type[_T] | None = ...) -> Callable[..., _R_co]: ...
    if sys.version_info >= (3, 10):
        __name__: str
        __qualname__: str
        @property
        def __wrapped__(self) -> Callable[..., _R_co]: ...
```

[PEP 673 - Self Type](https://peps.python.org/pep-0673/) describes `Self` as shorthand for: `TClassName = TypeVar("TClassName", bound="ClassName")`. So maybe covariant TypeVar would be the fix here?
Also PEP 673 was only just introducted in python 3.11. Maybe that could be a problem too?

Reply:
But then again djangos [classproperty](https://github.com/django/django/blob/0dd29209091280ccf34e07c9468746c396b7778e/django/utils/functional.py#L61-L75) is implemented a bit differently than CPythons [classmethod](https://github.com/python/cpython/blob/main/Objects/funcobject.c#L888)

Reply:
I just saw that `_Get` already is the covariant TypeVar, so that should be fine.
Looking at the [typehints for @property](https://github.com/python/typeshed/blob/307dd08a6e09f800cb239d37f0eb475345c7c67c/stdlib/builtins.pyi#L1168-L1185) I think I now understand what @sobolevn means (simply replace the remaining `Self` with `Any`):
```python
    fget: Callable[[Any], _Get] | None
    def __init__(self, method: Callable[[Any], _Get] | None = ...) -> None: ...
    def __get__(self, instance: Any | None, cls: type[Any] = ...) -> _Get: ...
    def getter(self, method: Callable[[Any], _Get]) -> classproperty[_Get]: ...
```

Reply:
Indeed works just as well with `Any`. I've pushed this change.